src/MoltenCampaign.sol

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: UNLICENSED
+// solhint-disable not-rely-on-time
+pragma solidity ^0.8.17;
+
+import {MToken} from "./MToken.sol";
+import {IERC20Votes} from "./interfaces/IERC20Votes.sol";
+
+contract MoltenElection {
+    IERC20Votes public daoToken;
+    // Threshold in daoToken-weis.
+    uint256 public threshold;
+    uint128 public duration;
+    uint128 public cooldownDuration;
+
+    constructor(
+        address daoTokenAddress,
+        uint256 _threshold,
+        uint128 _duration,
+        uint128 _cooldownDuration
+    ) {
+        daoToken = IERC20Votes(daoTokenAddress);
+        threshold = _threshold;
+        duration = _duration;
+        cooldownDuration = _cooldownDuration;
+    }
+
+    // 🎨 We could further dependency-inject by having a separate libs
+    // MTokenDeployer and MoltenCampaignDeployer which addresses are passed as
+    // argument of the constructor/function.
+    function makeCampaign(string calldata delegateName)
+        external
+        returns (MoltenCampaign)
+    {
+        MToken mToken = new MToken(
+            string.concat("Molten ", daoToken.name()),
+            string.concat("m", daoToken.symbol(), "-", delegateName),
+            address(this)
+        );
+        MoltenCampaign campaign = new MoltenCampaign(
+            msg.sender,
+            address(this),
+            address(mToken)
+        );
+        mToken.transferOwnership(address(campaign));
+        return campaign;
+    }
+}
+
+contract MoltenCampaign {
+    // Immutable props.
+    address public representative;
+    MoltenElection public election;
+    MToken public mToken;
```

Comment:
You can further enforce immutability of these system variables by adding the `immutable` modifier to these variables:

```solidity
    address public immutable representative;
    MoltenElection public immutable election;
    MToken public immutable mToken;
```

[also](https://docs.soliditylang.org/en/v0.8.17/contracts.html#constant-and-immutable-state-variables):
> Compared to regular state variables, the gas costs of constant and immutable variables are much lower.
> Immutable variables are evaluated once at construction time and their value is copied to all the places in the code where they are accessed.